### PR TITLE
Added bring to top/send to back

### PR DESCRIPTION
**Pull request by vitaliel**
Select/deselect shapes

### CODE REVIEW

**Comment by vitaliel** [comment]
@danschultz Please review

File: src/actions/index.ts [hunk 1]
```diff
@@ -7,7 +7,8 @@ export const addFigure = (type : string, x: number, y: number) => {
             id: nextFigureId++,
             type: type,
             x: x,
-            y: y
+            y: y,
+            selected: false
```
Comment (danschultz) [inline src/actions/index.ts]: i'd like to see a more defined model for your application. there's a canvas, canvas has elements of type shape rectangle, triangle, circle, etc.

File: src/actions/index.ts [hunk 2]
```diff
@@ -22,3 +23,30 @@ export const moveFigure = (id : number, x: number, y: number) => {
...
+    }
+};
+
+export const bringToTop = () => {
+    return {
+        type: 'BRING_TO_TOP'
+    }
+};
+
+export const deselectAll = () => {
```
Comment (danschultz) [inline src/actions/index.ts]: please use the function keyword when defining non-block functions.

``` ts
// good
export function deselectAll(): Action {
  return {type: 'deselect_all'};
}

// good
[1, 2, 3].map((value) => value + 1);

// bad
export const deselectAll = () => {
  ...
}

// bad
[1, 2, 3].map(function(value) {
  ...
});
```

File: src/components/FigureProps.tsx [hunk 3]
```diff
@@ -1,10 +1,8 @@
 export interface FigureProps {
     id: number;
     type: string;
-    width: number;
-    height: number;
     x: number;
     y: number;
+    selected?: boolean;
     draggable: boolean;
-    onMoveFigure?: (id: number, x: number, y :number) => void;
```
Comment (danschultz) [inline src/components/FigureProps.tsx:9]: the component's props is generally defined inside the component's module.

File: src/components/OnTopButton.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+class OnTopButton extends React.Component<any, any> {
+    render() {
+        return (
+            <button disabled={this.props.disabled} onClick={() => this.props.onClick()}>Bring to Top</button>
+        );
+    }
+}
+
+export default connect(mapStateToProps, mapDispatchToProps)(OnTopButton);
```
Comment (danschultz) [inline src/components/OnTopButton.tsx:28]: fyi, we'll be using snakecasing when naming files in the editor.

File: src/components/Shape.tsx [hunk 5]
```diff
@@ -52,23 +73,48 @@ class Shape extends React.Component<FigureProps, any> {
...
+
+        if (this.props.selected) {
+            this.dragElements = this.dragElements.concat('select_rect', 'p1', 'p2', 'p3', 'p4');
+
+            return <g>
+                <rect ref='select_rect' x={this.props.x} y={this.props.y} width={size} height={size} fill="#fff" stroke="#000" />
+                {this.renderSelectionPoint('p1', this.props.x, this.props.y)}
+                {this.renderSelectionPoint('p2', this.props.x+size, this.props.y)}
+                {this.renderSelectionPoint('p3', this.props.x, this.props.y+size)}
+                {this.renderSelectionPoint('p4', this.props.x+size, this.props.y+size)}
```
Comment (danschultz) [inline src/components/Shape.tsx]: i would recommend creating a separate component for the selection box, and putting it in it's own layer that's above all the other components. this way the user can see all the resize handles and has access to them even if there's a shape above the selected item.

screen shot 2016-07-14 at 10 50 34 am

Comment (vitaliel) [inline src/components/Shape.tsx]: @danschultz done

File: src/actions/index.ts (hunk 2 above)
Comment (vitaliel) [inline src/actions/index.ts]: @danschultz done

File: src/components/OnTopButton.tsx (hunk 4 above)
Comment (vitaliel) [inline src/components/OnTopButton.tsx:28]: ok

**Comment by vitaliel** [comment]
@danschultz please review

File: src/models/AppState.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                    return element;
+                }
+            });
+
+            return new AppState(new Canvas(newElements));
+        } else {
+            return this;
+        }
+    }
+}
```
Comment (danschultz) [inline src/models/AppState.ts:88]: 👍

File: src/components/ShapeSelection.tsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import * as React from "react";
+
+const size = 70;
+
+class ShapeSelection extends React.Component<any, any> {
```
Comment (danschultz) [inline src/components/ShapeSelection.tsx]: it's always a good idea to define a props interface. this way you get compile time checking for the component.

File: src/components/Shape.tsx [hunk 8]
```diff
@@ -39,17 +44,27 @@ class Shape extends React.Component<FigureProps & Dispatch, any> {
...
             (e: DragSessionEvent) => {
-                this.dragElements.forEach((ref:string, i:number) => {
-                    this.refs[ref].style.transform = "";
-                });
+                const selectedElements = document.querySelectorAll("#canvas_area .selected");
+
+                for (let i = 0; i < selectedElements.length; i++) {
+                    const element = selectedElements[i] as HTMLElement;
+                    element.style.transform = "";
+                }
```
Comment (danschultz) [inline src/components/Shape.tsx]: perhaps the drag move logic should be part of the selection overlay.

**Comment by danschultz** [comment]
some minor stuff. LGTM after changes

**Pull request merged**